ddtrace/settings.py

```diff
@@ -118,7 +168,17 @@ class IntegrationConfig(AttrDict):
         config.flask['service_name'] = 'my-service-name'
         config.flask.service_name = 'my-service-name'
     """
-    def __init__(self, global_config, *args, **kwargs):
+
+    # DEV: Need to explicitly declare the attributes as to ignore them when
```

Comment:
By having `IntegrationConfig` inherit `AttrDict`/`dict` we run into issues with having attributes that we want to be private/exclusive to the `IntegrationConfig` itself. This results in some nastiness in `__getattribute__`, `__getitem__` and `__setattr__`. And of course we have to be sure to keep this set up to date with whatever future attributes may be added to the `IntegrationConfig`.

To me it feels like we could solve this by having an `AttrDict` attribute within the `IntegrationConfig` that we use in `__getattribute__`, `__getitem__` and `__setattr__`. The usage of the `IntegrationConfig` would stay the same, but the implementation would be a lot cleaner, and I don't _think_ we sacrifice anything in the process. Thoughts on this design choice?

Reply:
We can leverage `object.__getattribute__` and `object.__setattr__` here.

It isn't the most fun, but works:

```python
from ddtrace.utils.attrdict import AttrDict


class Custom(AttrDict):
    def __init__(self, *args, **kwargs):
        super(Custom, self).__init__(*args, **kwargs)
        object.__setattr__(self, 'config', object())
        object.__setattr__(self, 'name', 'CustomName')

    def __getattribute__(self, key):
        try:
            return object.__getattribute__(self, key)
        except AttributeError:
            pass

        return self[key]

    def __setattr__(self, key, value):
        if hasattr(self, key):
            object.__setattr__(self, key, value)
        else:
            self[key] = value
```


```python
>>> c = Custom()
>>> c.__dict__
{'config': <object object at 0x7fa28a769150>, 'name': 'CustomName'}
>>> c.config
<object object at 0x7fa28a769150>
>>> c.name
'CustomName'
>>> c.text = 'value'
>>> c
{'text': 'value'}
>>>
```

Reply:
/shrug

no, probably ok to remove `AttrDict` from here.

As long as attribute and item accessors for configuration work, idc how we handle under the covers.

If you want a `self._data = dict()` to store/manage the actual configuration values rather than `IntegrationConfig` being a `dict` itself, fine by me.

Reply:
Cool, just wanted to make sure I wasn't missing anything. I'll check it out and see if it cleans things up.